docs/design-docs/08-package-variant.md

```diff
@@ -340,17 +375,77 @@ With PackageVariant, we instead create the following resources:
 
 ### PackageVariant API
 
-#### Basic API
+The Go types below defines the `PackageVariantSpec`.
 
-- Upstream
-- Downstream
-- Annotations
-- Labels
-- Adoption and Deletion Policy
-- Status
-  - `Valid` Condition Type
-  - `DownstreamEnsured` Condition Type (`Ready`?)
-  - `DraftExists` Condition Type (?)
+```go
+type PackageVariantSpec struct {
+        Upstream   *Upstream   `json:"upstream,omitempty"`
+        Downstream *Downstream `json:"downstream,omitempty"`
+
+        AdoptionPolicy AdoptionPolicy `json:"adoptionPolicy,omitempty"`
+        DeletionPolicy DeletionPolicy `json:"deletionPolicy,omitempty"`
+
+        Labels      map[string]string `json:"labels,omitempty"`
+        Annotations map[string]string `json:"annotations,omitempty"`
+
+        PackageContext *PackageContext     `json:"packageContext,omitempty"`
+        Pipeline       *kptfilev1.Pipeline `json:"pipeline,omitempty"`
+        Injectors      []InjectionSelector `json:"injectors,omitempty"`
+}
+
+type Upstream struct {
+        Repo     string `json:"repo,omitempty"`
+        Package  string `json:"package,omitempty"`
+        Revision string `json:"revision,omitempty"`
+}
+
+type Downstream struct {
+        Repo    string `json:"repo,omitempty"`
+        Package string `json:"package,omitempty"`
+}
+
+type PackageContext struct {
+        Data       map[string]string `json:"data,omitempty"`
+        RemoveKeys []string          `json:"removeKeys,omitempty"`
+}
+
+type InjectionSelector struct {
+        Group   *string `json:"group,omitempty"`
+        Version *string `json:"version,omitempty"`
+        Kind    *string `json:"kind,omitempty"`
+        Name    string  `json:"name"`
+}
+
+```
+
+#### Basic Spec Fields
+
+The `Upstream` and `Downstream` fields specify the source package and
+destination repository and package name. The `Repo` fields refer to the names
+Porch Repository resources in the same namespace as the PackageVariant resource.
+The `Downstream` does not contain a revision, because the package variant
+controller will only create Draft packages. The `Revision` of the eventual
+PackageRevision resource will be determined by Porch at the time of approval.
+
+The `Labels` and `Annotations` fields list metadata to include on the created
+PackageRevision. These values are set *only* at the time a Draft package is
+created. They are ignored for subsequent operations, even if the PackageRevision
```

Comment:
So this means that these are only added whenever a new draft PackageRevision is stamped out, but then users or other controllers are free to change these as they want to without the controller interfering right?

Reply:
Correct, I will make that clear.